net/xia/ppal_ether/main.c

```diff
@@ -351,6 +351,207 @@ static const xia_ppal_all_rt_eops_t ether_all_rt_eops = {
 	},
 };
 
+/* Routing process per principal struct. */
+static inline struct interface_addr *xdst_naddr(struct xip_dst *xdst)
+{
+	return xdst->info;
+}
+
+static int main_input_input(struct sk_buff *skb)
+{
+	struct xiphdr *xiph;
+	struct xip_dst *xdst;
+	struct interface_addr *naddr;
+
+	if (skb->pkt_type != PACKET_HOST)
+		goto drop;
+
+	xiph = xip_hdr(skb);
+	if (!xiph->hop_limit) {
+		net_warn_ratelimited("%s: hop limit reached\n", __func__);
+		goto drop;
+	}
+
+	xdst = skb_xdst(skb);
+
+	skb = xip_trim_packet_if_needed(skb, dst_mtu(&xdst->dst));
+	if (unlikely(!skb))
+		return NET_RX_DROP;
+
+	/* We are about to mangle packet. Copy it! */
+	naddr = xdst_naddr(xdst);
+	if (skb_cow(skb, LL_RESERVED_SPACE(naddr->outgress_interface) 
+		                    + xdst->dst.header_len))
+		goto drop;
+	xiph = xip_hdr(skb);
+
+	/* Decrease ttl after skb cow done. */
+	xiph->hop_limit--;
+
+	return dst_output(xdst_net(xdst), skb->sk, skb);
+
+drop:
+	kfree_skb(skb);
+	return NET_RX_DROP;
+}
+
+static inline struct interface_addr *skb_naddr(struct sk_buff *skb)
+{
+	return xdst_naddr(skb_xdst(skb));
+}
+
+static inline int xip_skb_dst_mtu(struct sk_buff *skb)
+{
+	return dst_mtu(skb_dst(skb));
+}
+
+static inline int neighinterface_hh_output(const struct hh_cache *hh, 
+            struct sk_buff *skb)
+{
+	unsigned int seq;
+	int hh_len;
+
+	do {
+		seq = read_seqbegin(&hh->hh_lock);
+		hh_len = hh->hh_len;
+		if (likely(hh_len <= HH_DATA_MOD)) 
+		{
+			memcpy(skb->data - HH_DATA_MOD, hh->hh_data, HH_DATA_MOD);
+		} 
+		else 
+		{
+			int hh_alen = HH_DATA_ALIGN(hh_len);
+			memcpy(skb->data - hh_alen, hh->hh_data, hh_alen);
+		}
+	} while (read_seqretry(&hh->hh_lock, seq));
+
+	skb_push(skb, hh_len);
+	return dev_queue_xmit(skb);
+}
+
+static inline int interface_neigh_output(struct fib_xid_ether_main *mfxid, 
+	        struct sk_buff *skb)
+{
+	const struct hh_cache *hh = &mfxid->cached_hdr;
+
+	if (hh->hh_len)
+		return neighinterface_hh_output(hh, skb);
+	else
+		return mfxid->output(mfxid, skb);
```

Comment:
In the final code,
1. field `output` of `struct fib_xid_ether_main` is always pointing to `mfxid_blackhole()` and
2. this is the only place that field `output` is used.

In other words, we don't need field `output`. We can just call `mfxid_blackhole()` here.

Moreover, `mfxid_blackhole()` is tiny and will be only used here after the changes above, so just merge `mfxid_blackhole()` here.

Finally, given the `if` statement above finishes with a `return`, we don't even need the `else` clause. This will reduce indentation.